README.md
# Ways to NOT make money
###A basic Python 2.x/3.x Cookie-Clicker style (incremental) game
##### patch-1 diverged from master (merging will create conflicts) - PLEASE KEEP BRANCHES UP-TO-DATE!

For the Python 2 version, go to the py2 folder, and run `game.py`.

For the Python 3 version, go to the py3 folder, and run `game.py`.

For the executable version (AKA "You're too lazy and/or you don't want to look at code"): go to the exe folder.

All folders contain the latest version of the game, **V0.7.1**.

IMPORTANT: Py2 and Py3 folders are pure Python. exe.0.7.1 and single-file-executable.0.7.1 are (almost) pure Tcl.


changelog.txt
Changelog
 0.7.2
        -Merged Py2 and Py3 together
        -
 0.7.1:
        -Added lotto system
 
 0.7.0:
        -MAJOR UPDATE!
        -Many new feautures were added, including,
           -Bank Heist (Building)
           -Multiplier
           -Bank Blueprints (Upgrade)
           -And more!
        - Fixed Tk threading issue.
 0.6.3b2:
        -Added in-game open browser for issue-reporting

 0.6.3:
        -Updates are rolling in!
        -Tweaked the savefile, so they don't constantly refresh
        -Changed the setup of the 'issues' box
        -IMPLEMENTED AUTOUPDATING for savefiles!

 0.6.2:
        -EVEN MORE MAJOR OF AN UPDATE!
        -Added multiplayer (you can play with your friends on the save device!)
            -Changed the mechanics of savegame()
        -Report issues in-game, you don't have to open web browser!

 0.6.1-alpha:
        -MAJOR UPDATE!
        -Improved the "cannot affords" and "no requirements", changing the sign to a label that appears in the same place instead of a pop-up window (because those are annoying).
        -Added "million" things to prices (which is great!)

 0.6.0-3:
        -Added bells to "cannot affords" and "no requirements".
        -Fixed more bugs (qlstudio edit this, I have no idea what else you fixed).

 0.6.0-2:
        -Fixed a variety of bugs (thanks qlstudio).
            -Fixed the OverflowError for the buildings themselves
            -Checked for conflict with all the "-illion"'s, no conflict (so far).

 0.6.0-1:
        -Further added to the "million", creating "billion" to "quintillion", which should delay the OverflowError further (most annoying error so far).
        -Fixed the "million", and now there should be no conflict.

 0.6.0:
        -Implemented a "millions" thing, which should prevent (or at least delay) the OverflowError

 0.5.3-3:
        -Fixed two main bugs, the timevar and totalclicksvar bugs.

 0.5.3-2:
        -Fixed a variety of bugs. (HOORAY FOR BUG FIXING)

 0.5.3-1:
        -Fixed the savefile (YAY)
        -Implemented an upgrades button
        -Started on stats

 0.5.2:
        -Added a colour changed on click of +Money Button (Yes, it's addictive)
        -Fixed Gold Button
        -Fixed a bug in which more than one gold would make a sticky button. (Only allowed one gold to exsist at a time)
        -Doubled Chance of finding Gold.
        -Working on a Stats Menu.

 0.5.1:
        -Fixed the upgrades button (it makes the main screen wider instead of creating a new window)
        -Starting work on background colours
        -Still fixing gold button (the previous update was a lie)

 0.5.0:
        -Probably fixed gold button
        -Added new upgrade for Money Printers (Costs $7,777,777)
        -Made a new upgrades button (less crowding)

 0.4.3:
        -Shrunk the code down by 100+ lines
        -Changed gold appearances to randint function
        -Fixed a bug in which upgrades don't change MPS

 0.4.2:
        -Added gold button (increases balance by MPS*50)
        -Fixed most of the bugs (PyCharm!)

 0.4.1:
        -Downloading images is no longer required (Thanks to Git Hub for Desktop)
        -Simple animation added. (To be improved)
        -Increased +Money Size (Easier for touchscreen users.
        -Bugs fixed (Prior to this upadate)
        -Made the change log easier to read. (Order Reversed)

 0.4.0:
        -Added new building, Sharemarket Crash (969 MPS, costs $42,000)
        -Changed the names of the upgrades (less bland)
        -Ironed out a LOT of the bugs
        -Changed the code for upgrades AGAIN
        -Added background

 0.3.0:
        -Added a new upgrade, Better Click Boost (Costs $200,000)
        -Changed the code for upgrades

 0.2.1 (not to be confused with 0.2-1): Fully fixed the savefile, and fixed bugs.

 0.2-2: Partially fixed the savefile.

 0.2-1*: Changed the save system so that it included an external savefile.

 0.2.0*: Added a save system (which needs fixing!), and a new building: Counterfeit Company (321 MPS, costs $9,001).

 0.1.3: Fixed a bug in which some labels and buttons don't update.

 0.1.2: Added placeholders for the Tkinter window.

 0.1.1: Changed from pack and grid to only grid, saving space and making the whole code work


game_model.py
import math


class GameState:
    def __init__(self, data=None):
        if data is None:
            self.data = [
                "auto", 0, "print", 0, "counter", 0, "shares", 0, "bank", 0, "upg1h1", 0, "upg1h2", 0,
                "upg2h1", 0, "upg2h2", 0, "upg3", 0, "upg4", 0, "upg5", 0, "cupg1", 0, "cupg2", 0,
                "money", 0.0, "time", 0, "clicks", 0, "lotto", 100]
        else:
            self.data = data

        print('GameState got data:', self.data)

        self.check = 0
        self.upgbuttoncheck = False
        self.goldcheck = False
        self.statscheck = False

        self.animate = 0
        self.clickcolourcheck = 1

        self.upgcheck1h1 = int(self.data[11])
        self.upgcheck1h2 = int(self.data[13])
        self.upgcheck2h1 = int(self.data[15])
        self.upgcheck2h2 = int(self.data[17])
        self.upgcheck3 = int(self.data[19])
        self.upgcheck4 = int(self.data[21])
        self.upgcheck5 = int(self.data[23])
        self.clickupgcheck1 = int(self.data[25])
        self.clickupgcheck2 = int(self.data[27])

        self.money = float(self.data[29])

        self.timeplay = int(self.data[31])
        self.totalclicks = int(self.data[33])
        self.lottoprice = int(self.data[35])

        self.autoclick2 = int(self.data[1])
        self.autoclick = self.autoclick2 * 18 * self.upgcheck1h2 + self.autoclick2 * 2 * self.upgcheck1h1 + \
                         self.autoclick2
        self.autoprice = float(20 * math.pow(1.15, self.autoclick2))

        self.printmoney2 = int(self.data[3])
        self.printmoney = self.printmoney2 * 18 * self.upgcheck2h2 + self.printmoney2 * 2 * self.upgcheck2h1 + \
                          self.printmoney2
        self.printprice = float(375 * math.pow(1.25, self.printmoney2))

        self.counterfeit2 = int(self.data[5])
        self.counterfeit = self.counterfeit2 * 2 * self.upgcheck3 + self.counterfeit2
        self.counterprice = float(9001 * math.pow(1.3, self.counterfeit2))

        self.sharecrash2 = int(self.data[7])
        self.sharecrash = self.sharecrash2 * 2 * self.upgcheck4 + self.sharecrash2
        self.shareprice = float(42000 * math.pow(1.4, self.sharecrash2))

        self.bankheist2 = int(self.data[9])
        self.bankheist = self.bankheist2 * 2 * self.upgcheck4 + self.bankheist2
        self.bankprice = float(175000 * math.pow(1.1, self.bankheist2))

        self.mps = self.autoclick2 + 15 * self.printmoney2 + 321 * self.counterfeit2 + 969 * self.sharecrash2 + \
                   2015 * self.bankheist2
        self.inc = 1 + self.clickupgcheck1 * 2 + self.clickupgcheck2 * self.mps / 10

        self.multiplier = 1

    def __str__(self):
        return str(self.__dict__)

    def can_increment(self, name):
        return self.money >= getattr(self, name)

    def increment(self, name):
        mult = self.multiplier

        if name == 'autoprice':
            print('buying auto')
            print('price:', self.autoprice)
            self.money -= self.autoprice
            self.autoclick += (1 + self.upgcheck1h1 * 2 + self.upgcheck1h2 * 18) * mult

            self.autoclick2 += mult
            print('self.autoclick2', self.autoclick2)

            self.mps += (1 + self.upgcheck1h1 * 2 + self.upgcheck1h2 * 18) * mult
            self.autoprice = 20 * (math.pow(1.15, self.autoclick2))
            self.inc += math.pow(self.clickupgcheck2 * self.autoclick, 1.01)

        elif name == 'printprice':
            self.money -= self.printprice
            self.printmoney += (1 + self.upgcheck2h1 * 2 + self.upgcheck2h2 * 18) * mult
            self.printmoney2 += mult
            self.mps += 15 * (1 + self.upgcheck2h1 * 2 + self.upgcheck2h2 * 18) * mult
            self.printprice = 375 * math.pow(1.25, self.printmoney2)
            self.inc += math.pow(self.clickupgcheck2 * self.printmoney, 1.01)

        elif name == 'counterprice':
            self.money -= self.counterprice
            self.counterfeit += (1 + self.upgcheck3 * 2) * mult
            self.counterfeit2 += mult
            self.mps += 221 * (1 + self.upgcheck3 * 2) * mult
            self.counterprice = 9001 * math.pow(1.3, self.counterfeit2)
            self.inc += math.pow(self.clickupgcheck2 * self.counterfeit, 1.01)

        elif name == 'shareprice':
            self.money -= self.shareprice
            self.sharecrash += (1 + self.upgcheck4 * 2) * mult
            self.sharecrash2 += mult
            self.mps += 969 * (1 + self.upgcheck4 * 2) * mult
            self.shareprice = 42000 * math.pow(1.4, self.sharecrash2)
            self.inc += math.pow(self.clickupgcheck2 * self.sharecrash, 1.01)

        elif name == 'bankprice':
            self.money -= self.bankprice
            self.bankheist += (1 + self.upgcheck4 * 2) * mult
            self.bankheist2 += mult
            self.mps += 1844 * (1 + self.upgcheck4 * 2) * mult
            self.bankprice = int(175000 * math.pow(1.1, self.bankheist2))
            self.inc += math.pow(self.clickupgcheck2 * self.bankheist, 1.01)

    def bump_multiplier(self):
        if self.multiplier == 100:
            self.multiplier = 1

        elif self.multiplier == 1:
            self.multiplier = 10

        elif self.multiplier == 10:
            self.multiplier = 100

        self.autoprice = 0
        self.printprice = 0
        self.counterprice = 0
        self.shareprice = 0
        self.bankprice = 0

        for a in range(self.multiplier):
            self.autoprice += float(20 * math.pow(1.15, (self.autoclick2 + a)))
            self.printprice += float(375 * math.pow(1.25, (self.printmoney2 + a)))
            self.counterprice += float(9001 * math.pow(1.3,  (self.counterfeit2 + a)))
            self.shareprice += float(42000 * math.pow(1.4,  (self.sharecrash2 + a)))
            self.bankprice += float(175000 * math.pow(1.1,  (self.bankheist2 + a)))

    def get_totalspent(self):
        x = 0
        for a in range(self.autoclick2):
            x += round(float(20 * math.pow(1.15, a)), 2)
        for b in range(self.printmoney2):
            x += round(float(375 * math.pow(1.25, b)), 2)
        for c in range(self.counterfeit2):
            x += round(float(9001 * math.pow(1.3, c)), 2)
        for d in range(self.sharecrash2):
            x += round(float(42000 * math.pow(1.4, d)), 2)
        for e in range(self.bankheist2):
            x += round(float(175000 * math.pow(1.1, e)), 2)
        return x
